baldur/src/main/java/org/opencadc/baldur/PermissionsConfig.java

```diff
@@ -272,45 +241,49 @@ static void clearCache() {
         entries = null;
     }
     
-    class PermissionEntry {
-        private String name;
-        private Pattern pattern;
-        boolean anonRead = false;
-        List<GroupURI> readOnlyGroups = new ArrayList<GroupURI>();
-        List<GroupURI> readWriteGroups = new ArrayList<GroupURI>();
+    /**
+     * Class that iterates over permission entries that match the
+     * given artifact URI.
+     */
+    class PermissionIterator implements Iterator<PermissionEntry> {
         
-        PermissionEntry(String name, Pattern pattern) {
-            this.name = name;
-            this.pattern = pattern;
-        }
+        URI artifactURI;
+        Iterator<PermissionEntry> entryIterator;
+        PermissionEntry next = null;
         
-        public String getName() {
-            return name;
+        PermissionIterator(URI artifactURI) {
+            this.artifactURI = artifactURI;
+            entryIterator = entries.iterator();
+            advance();
         }
-
-        public Pattern getPattern() {
-            return pattern;
+        
+        private void advance() {
+            if (entryIterator.hasNext()) {
+                PermissionEntry nextEntry = entryIterator.next();
+                Matcher matcher = nextEntry.getPattern().matcher(artifactURI.toString());
+                if (matcher.matches()) {
+                    next = nextEntry;
+                } else {
+                    advance();
+                }
+            } else {
+                next = null;
+            }
         }
-
-        @Override
-        public boolean equals(Object o) {
-            return this.name.equals(((PermissionEntry) o).getName());
+        
+        public boolean hasNext() {
+            return next != null;
         }
         
-        @Override
-        public String toString() {
-            StringBuilder sb = new StringBuilder("PermissionEntry=[");
-            sb.append("name=[").append(name).append("],");
-            sb.append("pattern=[").append(pattern).append("],");
-            sb.append("anonRead=[").append(anonRead).append("],");
-            sb.append("readOnlyGroups=[");
-            sb.append(Arrays.toString(readOnlyGroups.toArray()));
-            sb.append("],");
-            sb.append("readWriteGroups=[");
-            sb.append(Arrays.toString(readWriteGroups.toArray()));
-            sb.append("]");
-            return sb.toString();
+        public PermissionEntry next() {
+            if (next == null) {
```

Comment:
This should be NoSuchElementException (standard Iterator API); no message needed

I do like that the pattern matching iterator is now easy to grok.

Reply:
ok